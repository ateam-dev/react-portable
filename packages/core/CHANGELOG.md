# @react-portable/core

## 0.7.0-next.10

### Patch Changes

- [#127](https://github.com/ateam-dev/react-portable/pull/127) [`1804dfa`](https://github.com/ateam-dev/react-portable/commit/1804dfa8e6fb18d5ce43714c22bfc4cac2404540) Thanks [@aiji42](https://github.com/aiji42)! - Optimize for better stability

- Updated dependencies [[`1804dfa`](https://github.com/ateam-dev/react-portable/commit/1804dfa8e6fb18d5ce43714c22bfc4cac2404540)]:
  - @react-portable/gateway@0.5.0-next.8

## 0.7.0-next.9

### Patch Changes

- [#123](https://github.com/ateam-dev/react-portable/pull/123) [`77e1ea4`](https://github.com/ateam-dev/react-portable/commit/77e1ea41ab7c5634662f68221bc302f598ac0e3d) Thanks [@aiji42](https://github.com/aiji42)! - upgrade qwik version 1.2.8

- [`e49ccb6`](https://github.com/ateam-dev/react-portable/commit/e49ccb6aff5e111ae07035a43176a2285d02c72e) Thanks [@aiji42](https://github.com/aiji42)! - update peerDependencies qwik libraries

- [#122](https://github.com/ateam-dev/react-portable/pull/122) [`a67179a`](https://github.com/ateam-dev/react-portable/commit/a67179a2ae0d7d200c788dcdd838b337291077e6) Thanks [@aiji42](https://github.com/aiji42)! - fix qwik dependencies verson

- Updated dependencies [[`e17a3cf`](https://github.com/ateam-dev/react-portable/commit/e17a3cfefe76fbc9b0a0095dec9f068e6b6e28f3), [`a67179a`](https://github.com/ateam-dev/react-portable/commit/a67179a2ae0d7d200c788dcdd838b337291077e6)]:
  - @react-portable/gateway@0.5.0-next.7

## 0.7.0-next.8

### Minor Changes

- [#114](https://github.com/ateam-dev/react-portable/pull/114) [`2c03663`](https://github.com/ateam-dev/react-portable/commit/2c036632ecfc01827f5ee3fcc33f09eb7c80de87) Thanks [@aiji42](https://github.com/aiji42)! - Add support for hot reloading

### Patch Changes

- Updated dependencies [[`2c03663`](https://github.com/ateam-dev/react-portable/commit/2c036632ecfc01827f5ee3fcc33f09eb7c80de87)]:
  - @react-portable/gateway@0.5.0-next.6

## 0.7.0-next.7

### Minor Changes

- [#111](https://github.com/ateam-dev/react-portable/pull/111) [`b7a3142`](https://github.com/ateam-dev/react-portable/commit/b7a3142df26d30a6ba0275df1798d65647c75db3) Thanks [@aiji42](https://github.com/aiji42)! - Enable automatic loading of previewify.config.ts

## 0.7.0-next.6

### Minor Changes

- [#110](https://github.com/ateam-dev/react-portable/pull/110) [`63e1780`](https://github.com/ateam-dev/react-portable/commit/63e17803701284e8bfa3eb4915d4000122dc8122) Thanks [@aiji42](https://github.com/aiji42)! - the CLI commands have been improved: portable prepare is no longer needed and is completed with a single portable previewify.

- [#107](https://github.com/ateam-dev/react-portable/pull/107) [`b305d76`](https://github.com/ateam-dev/react-portable/commit/b305d76900ff4b53bf647bb89e21ac128475018d) Thanks [@aiji42](https://github.com/aiji42)! - 'remove features for portable'

### Patch Changes

- Updated dependencies [[`b305d76`](https://github.com/ateam-dev/react-portable/commit/b305d76900ff4b53bf647bb89e21ac128475018d)]:
  - @react-portable/gateway@0.5.0-next.5

## 0.7.0-next.5

### Minor Changes

- [#102](https://github.com/ateam-dev/react-portable/pull/102) [`4486e6c`](https://github.com/ateam-dev/react-portable/commit/4486e6c8ddfb0a2eaa8447a3dd620f28ca5a6807) Thanks [@aiji42](https://github.com/aiji42)! - support rendering react elements passed by props

### Patch Changes

- [#100](https://github.com/ateam-dev/react-portable/pull/100) [`c845c6d`](https://github.com/ateam-dev/react-portable/commit/c845c6dfbd862af50b3b8ff8a4358de8933859f5) Thanks [@aiji42](https://github.com/aiji42)! - prepare for supporting to preview ReactElement props

- Updated dependencies [[`4486e6c`](https://github.com/ateam-dev/react-portable/commit/4486e6c8ddfb0a2eaa8447a3dd620f28ca5a6807)]:
  - @react-portable/gateway@0.5.0-next.3

## 0.7.0-next.4

### Patch Changes

- [#97](https://github.com/ateam-dev/react-portable/pull/97) [`86a43ad`](https://github.com/ateam-dev/react-portable/commit/86a43adcd8d6443f2efebe2c74dd2538998cfb65) Thanks [@aiji42](https://github.com/aiji42)! - fix preview worker config when restarting

## 0.7.0-next.3

### Patch Changes

- [#95](https://github.com/ateam-dev/react-portable/pull/95) [`be9055f`](https://github.com/ateam-dev/react-portable/commit/be9055fcff8b38f827cdb6da3236235faad825d6) Thanks [@aiji42](https://github.com/aiji42)! - stabilize build task

## 0.7.0-next.2

### Patch Changes

- [#94](https://github.com/ateam-dev/react-portable/pull/94) [`0e59614`](https://github.com/ateam-dev/react-portable/commit/0e59614807f7d55dbd98f33942d1f9f49e4381f4) Thanks [@aiji42](https://github.com/aiji42)! - remove rp-preview included in fragment

- [#91](https://github.com/ateam-dev/react-portable/pull/91) [`b8e336b`](https://github.com/ateam-dev/react-portable/commit/b8e336b4f5717b4c69f1aaf0e82e3836e78959eb) Thanks [@aiji42](https://github.com/aiji42)! - beautify cli logs

- Updated dependencies [[`0e59614`](https://github.com/ateam-dev/react-portable/commit/0e59614807f7d55dbd98f33942d1f9f49e4381f4)]:
  - @react-portable/gateway@0.5.0-next.2

## 0.7.0-next.1

### Minor Changes

- [#81](https://github.com/ateam-dev/react-portable/pull/81) [`5c5c595`](https://github.com/ateam-dev/react-portable/commit/5c5c595a90ecb0db06d487be8b69b0eb2b956eab) Thanks [@aiji42](https://github.com/aiji42)! - start and integrate the gateway and component servers for preview with a single command

### Patch Changes

- Updated dependencies [[`5c5c595`](https://github.com/ateam-dev/react-portable/commit/5c5c595a90ecb0db06d487be8b69b0eb2b956eab)]:
  - @react-portable/gateway@0.5.0-next.1

## 0.7.0-next.0

### Minor Changes

- [#76](https://github.com/ateam-dev/react-portable/pull/76) [`c3f2d4d`](https://github.com/ateam-dev/react-portable/commit/c3f2d4d8112fd8475b83d84245a88c7a4e6650c6) Thanks [@aiji42](https://github.com/aiji42)! - rename modules and web components for shortify

- [#77](https://github.com/ateam-dev/react-portable/pull/77) [`4635d9e`](https://github.com/ateam-dev/react-portable/commit/4635d9e5cedb6a417c83fdc453c32fa1832262a9) Thanks [@aiji42](https://github.com/aiji42)! - export `previewify` as alias of `portable`

- [#78](https://github.com/ateam-dev/react-portable/pull/78) [`992c4a6`](https://github.com/ateam-dev/react-portable/commit/992c4a616d794c0432ba6db3f1ddddecaf141831) Thanks [@aiji42](https://github.com/aiji42)! - separate vite plugin and cli commands for build and preparation

### Patch Changes

- [#79](https://github.com/ateam-dev/react-portable/pull/79) [`4bb0290`](https://github.com/ateam-dev/react-portable/commit/4bb0290f0d7aee0902c346cd768c330dd04de1ca) Thanks [@aiji42](https://github.com/aiji42)! - fix support `previewify`

## 0.6.0

### Minor Changes

- [#59](https://github.com/ateam-dev/react-portable/pull/59) [`798dfe7`](https://github.com/ateam-dev/react-portable/commit/798dfe7ba03612cad6124a7296e8fdb440e4a17f) Thanks [@aiji42](https://github.com/aiji42)! - support for preview of children and functional props

## 0.5.0

### Minor Changes

- [#54](https://github.com/ateam-dev/react-portable/pull/54) [`5c70f44`](https://github.com/ateam-dev/react-portable/commit/5c70f44ee445aa686bfecb8eb857f676e23b9c25) Thanks [@aiji42](https://github.com/aiji42)! - import css on building

## 0.4.1

### Patch Changes

- [#51](https://github.com/ateam-dev/react-portable/pull/51) [`8c56847`](https://github.com/ateam-dev/react-portable/commit/8c5684777ee108c7b8ee8a58128305a047b5c34a) Thanks [@aiji42](https://github.com/aiji42)! - sanitizing portable code and fix dependencies of qwik libraries

## 0.4.0

### Minor Changes

- [#47](https://github.com/ateam-dev/react-portable/pull/47) [`5d32caf`](https://github.com/ateam-dev/react-portable/commit/5d32cafc27d80b99584604b15cff2c38e526c439) Thanks [@aiji42](https://github.com/aiji42)! - make manual installation of Qwik libraries unnecessary

## 0.3.1

### Patch Changes

- [#43](https://github.com/ateam-dev/react-portable/pull/43) [`72048f5`](https://github.com/ateam-dev/react-portable/commit/72048f55022be5f79d95dee3be87bae3344a46bc) Thanks [@aiji42](https://github.com/aiji42)! - bumpup and remove `@builder.io/qwik` and `@builder.io/qwik-city` from core dependencies

- [#41](https://github.com/ateam-dev/react-portable/pull/41) [`4b6a803`](https://github.com/ateam-dev/react-portable/commit/4b6a8032cdf724302bf89688d4c62c02b96fe48e) Thanks [@aiji42](https://github.com/aiji42)! - fix cli option description

## 0.3.0

### Minor Changes

- [#38](https://github.com/ateam-dev/react-portable/pull/38) [`80b57ce`](https://github.com/ateam-dev/react-portable/commit/80b57ce375318a1e8a229f4f55fd3902358f8c80) Thanks [@aiji42](https://github.com/aiji42)! - add feature for preview components on deployed page

## 0.2.0

### Minor Changes

- [#25](https://github.com/ateam-dev/react-portable/pull/25) [`3190319`](https://github.com/ateam-dev/react-portable/commit/3190319d3c0cde8c8344a7f5f3cf508106a76d6c) Thanks [@aiji42](https://github.com/aiji42)! - Support error handling within data loader

## 0.1.1

### Patch Changes

- [#17](https://github.com/ateam-dev/react-portable/pull/17) [`d953a9f`](https://github.com/ateam-dev/react-portable/commit/d953a9f4b9fb8e40a4f291af444e8a872ca7a5bf) Thanks [@aiji42](https://github.com/aiji42)! - bumpup

## 0.1.0

### Minor Changes

- [#14](https://github.com/ateam-dev/react-portable/pull/14) [`e4eb42f`](https://github.com/ateam-dev/react-portable/commit/e4eb42f1df4fca90d312e5c93251d1a7ea95f1d4) Thanks [@aiji42](https://github.com/aiji42)! - remove unnecessary cli options (@react-portable/core)

## 0.0.2

### Patch Changes

- [#3](https://github.com/ateam-dev/react-portable/pull/3) [`d073b81`](https://github.com/ateam-dev/react-portable/commit/d073b8196680285c85bf70a5484df4fda7c8dff3) Thanks [@aiji42](https://github.com/aiji42)! - first release!

## 0.0.1

### Patch Changes

- [#1](https://github.com/ateam-dev/react-portable/pull/1) [`9e7866c`](https://github.com/ateam-dev/react-portable/commit/9e7866c26b7ca994a5d32a177759f66bafe3800c) Thanks [@aiji42](https://github.com/aiji42)! - first release
