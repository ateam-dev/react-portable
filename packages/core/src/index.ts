export { previewify } from "./previewify";
export type { PreviewifyComponent } from "./previewify";
