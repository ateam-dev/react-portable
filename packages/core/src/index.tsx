import { qwikify$ } from "@builder.io/qwik-react";
import { component$ } from "@builder.io/qwik";
import { RequestHandler, routeLoader$ } from "@builder.io/qwik-city";
import * as Entry from "react-portable:virtual";

const { hydrate, revalidate = 0 } = Entry.strategy ?? {};

const qwikifyOption =
  hydrate === "onUse"
    ? ({
        eagerness: "hover",
        event: "focusin",
      } as const)
    : hydrate === "onIdle"
    ? ({
        eagerness: "idle",
        event: undefined,
      } as const)
    : {
        eagerness: undefined,
        event: undefined,
      };
const QComponent = qwikify$(Entry.default, qwikifyOption);
const getProps = routeLoader$(({ request }) => {
  return Entry.loader?.(request) ?? {};
});
export default component$(() => {
  const props = getProps().value;
  return <QComponent {...props} />;
});

export const onRequest: RequestHandler = async (requestEvent) => {
  requestEvent.headers.set(
    "Cache-Control",
    revalidate === false
      ? `public, s-maxage=${3600 * 24 * 365}`
      : revalidate === 0
      ? `no-store`
      : `public, s-maxage=${revalidate}, stale-while-revalidate=${
          3600 * 24 * 365
        }`
  );
};
