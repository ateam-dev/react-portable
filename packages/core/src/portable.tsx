import React, { FunctionComponent, useEffect, useRef } from "react";
import { RpPreview } from "@react-portable/client/web-components";
import { RequestEventCommon } from "@builder.io/qwik-city/middleware/request-handler";

export type Strategy = {
  revalidate?: false | 0 | number;
  hydrate?: "onUse" | "onIdle" | "disable";
};

type ErrorResponse = ReturnType<RequestEventCommon["error"]>;

export type Loader<T extends Record<string, unknown> = {}> = (
  r: Request,
  ctx: { error: RequestEventCommon["error"] },
) => T | ErrorResponse | Promise<T | ErrorResponse>;

export interface PortableComponent<T extends Record<string, unknown> = {}>
  extends FunctionComponent<T> {
  __code: string;
  __strategy: undefined | Strategy;
  __loader: undefined | Loader<T>;
  __original: FunctionComponent<T>;
}

type InferProps<T> = T extends FunctionComponent<infer U>
  ? U
  : T extends (props: infer U) => JSX.Element
  ? U
  : T extends () => JSX.Element
  ? {}
  : never;

type PortableOption<
  T extends FunctionComponent | ((props: any) => JSX.Element),
> = {
  loader?: Loader<InferProps<T>>;
  strategy?: Strategy;
  disablePreview?: boolean;
};

export const portable = <
  T extends FunctionComponent | ((props: any) => JSX.Element),
>(
  Component: T,
  code: string,
  { loader, strategy, disablePreview = false }: PortableOption<T> = {},
): PortableComponent<InferProps<T>> => {
  const Wrapped = (props: InferProps<T>) => {
    const ref = useRef<RpPreview>(null);
    useEffect(() => {
      if (!ref.current) return;

      ref.current.props = props;
      ref.current.rerender?.();
    }, [props]);

    if (disablePreview) return <Component {...(props as any)} />;

    return (
      <rp-preview ref={ref} code={code}>
        <Component
          {...(props as any)}
          children={<rp-outlet>{props.children}</rp-outlet>}
        />
      </rp-preview>
    );
  };

  Wrapped.__code = code;
  Wrapped.__loader = loader;
  Wrapped.__strategy = strategy;
  Wrapped.__original = Component;

  return Wrapped;
};

export const previewify = <
  T extends FunctionComponent | ((props: any) => JSX.Element),
>(
  Component: T,
  code: string,
): PortableComponent<InferProps<T>> => {
  return portable(Component, code, { strategy: { hydrate: "onIdle" } });
};
