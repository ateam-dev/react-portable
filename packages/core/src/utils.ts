export const currentDir = () => __dirname;
