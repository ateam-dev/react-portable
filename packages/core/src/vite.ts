import * as fs from "node:fs/promises";
import * as path from "node:path";
import * as ts from "typescript";
import type { PluginOption } from "vite";
import * as appRootPath from "app-root-path";
import { qwikCity } from "@builder.io/qwik-city/vite";
import { qwikVite } from "@builder.io/qwik/optimizer";
import { qwikReact } from "@builder.io/qwik-react/vite";
import { currentDir } from "./utils";

const copyAndReplace = async (
  srcPath: string,
  destPath: string,
  replacements: {
    search: string;
    replace: string;
  }[] = [],
): Promise<void> => {
  await fs.mkdir(path.dirname(destPath), { recursive: true });

  let result = await fs.readFile(srcPath, "utf-8");
  for (const { search, replace } of replacements) {
    result = result.replace(new RegExp(search, "g"), replace);
  }

  await fs.writeFile(destPath, result, "utf-8");
};

const findPortableFunctionCalls = (node: ts.Node): string | null => {
  if (
    ts.isCallExpression(node) &&
    ts.isIdentifier(node.expression) &&
    typeof node.expression.escapedText === "string" &&
    ["previewify"].includes(node.expression.escapedText) &&
    node.arguments.length >= 2
  ) {
    const [, secondArg] = node.arguments;

    if (ts.isStringLiteral(secondArg)) return secondArg.text;
  }

  let result = null;
  ts.forEachChild(node, (child) => {
    const childResult = findPortableFunctionCalls(child);
    if (childResult) result = childResult;
  });

  return result;
};

const getPortableCode = async (fileName: string) => {
  const script = await fs.readFile(fileName, "utf-8");
  const sourceFile = ts.createSourceFile(
    "test.ts",
    script,
    ts.ScriptTarget.Latest,
    true,
  );

  if (sourceFile) return findPortableFunctionCalls(sourceFile);
};

export const preparePlugin = (): PluginOption => {
  return {
    name: "react-portable-prepare",
    enforce: "pre",
    config: async (config) => {
      await Promise.all(
        ["root.tsx", "entry.ssr.tsx", "worker.ts"].map((file) => {
          return copyAndReplace(
            path.resolve(currentDir(), "../src/templates", file),
            path.resolve(portableConfig.coreDir, file),
          );
        }),
      );
      return config;
    },
    resolveId: async (source, importer) => {
      if (importer && source === "@react-portable/core") {
        const code = await getPortableCode(importer);
        if (code) {
          const destPath = path.resolve(
            portableConfig.coreDir,
            "routes",
            code,
            "index.tsx",
          );
          await copyAndReplace(
            path.resolve(currentDir(), "../src/templates", "route.tsx"),
            destPath,
            [
              { search: "__code__", replace: code },
              {
                search: "__sanitized__",
                replace: code.replace(/[-/:;*]/g, "_"),
              },
              { search: "__entryPath__", replace: importer },
              { search: "../portable", replace: "@react-portable/core" },
            ],
          );
        }
      }
    },
  };
};

type Config = {
  css: string | undefined;
  outDir: string;
  coreDir: string;
  entry: string;
};
const initialConfig: Config = {
  css: undefined,
  outDir: ".portable",
  coreDir: path.resolve(appRootPath.path, "node_modules", ".portable"),
  entry: "./src",
};
export let portableConfig = initialConfig;

export const previewifyPlugin = ({
  css,
  outDir,
  entry,
}: Partial<Config> = {}): PluginOption => {
  if (css) portableConfig.css = css;
  if (outDir) portableConfig.outDir = outDir;
  if (entry) portableConfig.entry = entry;

  return [];
};

export const resetConfig = () => (portableConfig = initialConfig);

export const portablePlugin = (): PluginOption => {
  return [
    {
      name: "react-portable-build",
      enforce: "pre",
      config: (config, env) => {
        if (env.ssrBuild) {
          return {
            build: {
              rollupOptions: {
                input: [
                  path.resolve(portableConfig.coreDir, "worker.ts"),
                  "@qwik-city-plan",
                ],
              },
            },
          };
        }
        return config;
      },
      transform: (code: string, id: string) => {
        if (
          id === path.resolve(portableConfig.coreDir, "root.tsx") &&
          portableConfig.css
        ) {
          return `${code}\nimport '${path.resolve(portableConfig.css)}'`;
        }

        if (
          !id.includes(portableConfig.coreDir) &&
          (id.endsWith(".tsx") || id.endsWith(".jsx"))
        ) {
          return `/** @jsxImportSource react */\n${code}`;
        }
      },
    },
  ];
};

export const qwikPlugins = (): PluginOption => {
  return [
    qwikCity({
      routesDir: path.resolve(portableConfig.coreDir, "routes"),
      trailingSlash: false,
    }),
    qwikVite({
      srcDir: portableConfig.coreDir,
      client: {
        outDir: path.resolve(portableConfig.outDir, "client"),
      },
      ssr: {
        outDir: path.resolve(portableConfig.outDir, "server"),
      },
      // vendorRoots: [
      //   path.dirname(require.resolve("@builder.io/qwik-city")),
      //   path.dirname(require.resolve("@builder.io/qwik-react")),
      // ],
    }),
    qwikReact(),
  ];
};
