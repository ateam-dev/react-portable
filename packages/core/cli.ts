#!/usr/bin/env node
import * as fs from "fs";
import * as path from "path";
import * as vite from "vite";
import { qwikVite } from "@builder.io/qwik/optimizer";
import { qwikCity } from "@builder.io/qwik-city/vite";
import tsconfigPaths from "vite-tsconfig-paths";
import { qwikReact } from "@builder.io/qwik-react/vite";
import { program } from "commander";
import { PluginOption } from "vite";
import fsx from "fs-extra";
import chokidar from "chokidar";
import { glob } from "glob";
import { Config } from "./types";

// FIXME: projectRoot
const projectRoot = process.cwd();
const baseModuleDir = path.resolve(__dirname, "../src");
let srcDir: string;
let clientOutDir: string;
let serverOutDir: string;
let tmpDir: string;
let entryFile: string;
let workerFilePath: string;

const loadConfig = () => {
  const config: Config = require(path.resolve(
    process.cwd(),
    "react-portable.config.js"
  ));

  srcDir = config.src;
  tmpDir = config.tmpDir ?? path.resolve(projectRoot, ".rp/tmp");
  clientOutDir =
    config.outDirs?.client ?? path.resolve(projectRoot, ".rp/client");
  serverOutDir =
    config.outDirs?.server ?? path.resolve(projectRoot, ".rp/server");
  workerFilePath =
    config.workerScript ?? path.resolve(projectRoot, "worker.ts");
  entryFile = path.resolve(tmpDir, "entry.ssr.tsx");
};

const vitePlugins = (): PluginOption[] => [
  {
    enforce: "pre",
    name: "react-portable-vite",
    resolveId: async (source) => {
      if (source.startsWith("react-portable:virtual:")) {
        const name = source.replace("react-portable:virtual:", "");
        const [target] = await glob(
          `${srcDir}/**/${name}.rp.@(ts|tsx|js|jsx)`,
          {
            absolute: true,
          }
        );
        return target;
      }
    },
    transform: (code: string, id: string) => {
      if (
        !id.includes(tmpDir) &&
        (id.endsWith(".tsx") || id.endsWith(".jsx"))
      ) {
        return `/** @jsxImportSource react */\n${code}`;
      }
    },
  },
  {
    enforce: "pre",
    name: "react-portable-worker",
    resolveId: (source) => {
      if (source === "@entry") return entryFile;
    },
  },
  qwikCity({
    routesDir: path.resolve(tmpDir, "routes"),
    trailingSlash: false,
  }),
  qwikVite({
    srcDir: tmpDir,
    client: {
      outDir: clientOutDir,
    },
    ssr: {
      outDir: serverOutDir,
    },
  }),
  tsconfigPaths(),
  qwikReact(),
];

const initProject = async () => {
  await Promise.all(
    ["react-portable.config.js", "worker.ts", "wrangler.toml"].map(
      async (file) => {
        // FIXME: check existing
        await fsx.promises.cp(
          path.resolve(baseModuleDir, file),
          path.resolve(projectRoot, file)
        );
        console.log(`ðŸ§© Placed ${file}`);
      }
    )
  );
};

const prepareProject = async () => {
  if (fs.existsSync(tmpDir)) fs.rmSync(tmpDir, { recursive: true });
  await Promise.all(
    ["root.tsx", "entry.ssr.tsx"].map((file) =>
      fsx.promises.cp(
        path.resolve(baseModuleDir, file),
        path.resolve(tmpDir, file)
      )
    )
  );
};

const syncRoutes = async (once: boolean = false) => {
  const routeFileTemplate = fsx.readFileSync(
    path.resolve(baseModuleDir, "index.tsx"),
    "utf8"
  );

  const watcher = chokidar.watch(`${srcDir}/**/*.rp.@(ts|tsx|js|jsx)`, {
    persistent: !once,
    ignoreInitial: false,
    followSymlinks: true,
    depth: 99,
  });

  watcher.on("unlink", (filePath) => {
    const name = filePath.match(/([^/]+?)\.rp\.[jt]sx?$/)![1];

    fsx.rmSync(path.resolve(tmpDir, "routes", ...name.split("."), "index.tsx"));
  });

  watcher.on("add", (filePath) => {
    const name = filePath.match(/([^/]+?)\.rp\.[jt]sx?$/)![1];

    fsx.outputFile(
      path.resolve(tmpDir, "routes", ...name.split("."), "index.tsx"),
      routeFileTemplate.replace(
        "react-portable:virtual",
        `react-portable:virtual:${name}`
      )
    );
  });

  return new Promise((r) => watcher.on("ready", r));
};

const serveSSR = async (option: { port?: number } = {}) => {
  const server = await vite.createServer({
    plugins: vitePlugins(),
    mode: "ssr",
  });
  await server.listen(option.port);
  server.printUrls();
};

const buildClient = async () => {
  return vite.build({
    plugins: vitePlugins(),
  });
};

const buildWorker = async (seamless = false) => {
  return vite.build({
    plugins: vitePlugins(),
    build: {
      emptyOutDir: !seamless,
      ssr: true,
      rollupOptions: {
        input: [workerFilePath, "@qwik-city-plan"],
      },
    },
  });
};

program.version("0.0.1");

program
  .command("init")
  .description("initialize project")
  .action(async () => {
    await initProject();
    console.log("ðŸŽ’Completed initializing react portable project");
  });

program
  .command("dev")
  .description("launch a server for development")
  .option("-p, --port <number>", "port number")
  .action(async ({ port }: { port?: string }) => {
    loadConfig();
    await prepareProject();
    await syncRoutes();

    await serveSSR({ port: port ? Number(port) : undefined }).catch((e) => {
      console.error(e);
      process.exit(1);
    });
  });

program
  .command("build")
  .description("build scripts for react portable")
  .action(async () => {
    loadConfig();
    await prepareProject();
    await syncRoutes(true);
    await buildClient();
    await buildWorker();
  });

program
  .command("watch")
  .description("watch mode")
  .option(
    "--preBuilt",
    "skip out-directory reset and initial build if already pre-built"
  )
  .action(async ({ preBuilt }: { watch?: true; preBuilt?: true }) => {
    loadConfig();
    if (!preBuilt) await prepareProject();
    await syncRoutes();
    if (!preBuilt) await buildClient();
    if (!preBuilt) await buildWorker();

    const watcher = chokidar.watch(`${srcDir}/**/*`, {
      persistent: true,
      ignoreInitial: true,
      depth: 99,
    });
    watcher.on("all", async () => {
      await buildClient();
      await buildWorker(true);
    });
  });

program.parse(process.argv);

if (program.args.length === 0) {
  program.help();
}
