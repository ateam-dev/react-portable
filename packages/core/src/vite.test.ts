import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import { reactPortablePlugin } from "./vite";
import { ConfigEnv, Plugin, UserConfig } from "vite";
import { fs, vol } from "memfs";
import * as fsPro from "node:fs/promises";
import * as utils from "./utils";

vi.mock("node:fs/promises");
vi.mock("./utils");

vi.mock("app-root-path", () => ({
  path: "/working-dir",
}));

type ConfigFunction = (
  config: UserConfig,
  env?: ConfigEnv,
) => Promise<UserConfig>;

type TransformFunction = (code: string, id: string) => string;

type ResolveIdFunction = (source: string, importer?: string) => Promise<void>;

const sampleComponentFile = `
import { portable } from '@react-portable/core'
const Component = () => <div></div>
export const Sample = portable(Component, 'sample/foo-bar')`;

const routeFileDummy = `
import { PortableComponent } from "../portable";
import * as Entries from "__entryPath__";
const Entry = Object.values(Entries).find((module) => {
  if (typeof module === "function" && "__code" in module)
    return module.__code === "__code__";
}) as PortableComponent;
const QComponent__sanitized__ = qwikify$(Entry, qwikifyOption);`;

describe("reactPortablePlugin", () => {
  describe("react-portable-prepare", () => {
    const plugin = reactPortablePlugin({ prepare: true }) as Plugin;
    beforeEach(() => {
      vol.fromJSON({
        "/working-dir/node_modules/@react-portable/core/src/templates/root.tsx":
          "console.log('root.tsx')",
        "/working-dir/node_modules/@react-portable/core/src/templates/entry.ssr.tsx":
          "console.log('entry.ssr.tsx')",
        "/working-dir/node_modules/@react-portable/core/src/templates/worker.ts":
          "console.log('worker.ts')",
        "/working-dir/node_modules/@react-portable/core/src/templates/route.tsx":
          routeFileDummy,
        "/working-dir/src/components/sample.tsx": sampleComponentFile,
      });
      // @ts-ignore
      vi.spyOn(fsPro, "mkdir").mockImplementation(fs.promises.mkdir);
      // @ts-ignore
      vi.spyOn(fsPro, "readFile").mockImplementation(fs.promises.readFile);
      // @ts-ignore
      vi.spyOn(fsPro, "writeFile").mockImplementation(fs.promises.writeFile);

      vi.spyOn(utils, "currentDir").mockReturnValue(
        "/working-dir/node_modules/@react-portable/core/dist",
      );
    });
    afterEach(() => {
      vol.reset();
    });

    test("`name` and `enforce`", async () => {
      expect(plugin.name).toBe("react-portable-prepare");
      expect(plugin.enforce).toBe("pre");
    });

    test("when `config` will be called, some files will be placed", async () => {
      await (plugin.config as ConfigFunction)({});

      expect(vol.toJSON("/working-dir/node_modules/.portable")).toStrictEqual({
        "/working-dir/node_modules/.portable/entry.ssr.tsx":
          "console.log('entry.ssr.tsx')",
        "/working-dir/node_modules/.portable/root.tsx":
          "console.log('root.tsx')",
        "/working-dir/node_modules/.portable/worker.ts":
          "console.log('worker.ts')",
      });
    });

    test("Install the route file if `portable` is called in the file", async () => {
      await (plugin.resolveId as ResolveIdFunction)(
        "@react-portable/core",
        "/working-dir/src/components/sample.tsx",
      );

      expect(
        vol.toJSON(
          "/working-dir/node_modules/.portable/routes/sample/foo-bar/index.tsx",
        ),
      ).toStrictEqual({
        "/working-dir/node_modules/.portable/routes/sample/foo-bar/index.tsx": `
import { PortableComponent } from "@react-portable/core";
import * as Entries from "/working-dir/src/components/sample.tsx";
const Entry = Object.values(Entries).find((module) => {
  if (typeof module === "function" && "__code" in module)
    return module.__code === "sample/foo-bar";
}) as PortableComponent;
const QComponentsample_foo_bar = qwikify$(Entry, qwikifyOption);`,
      });
    });
  });

  describe("react-portable-build", () => {
    const [plugin] = reactPortablePlugin() as [Plugin];

    test("`name` and `enforce`", () => {
      expect(plugin.name).toBe("react-portable-build");
      expect(plugin.enforce).toBe("pre");
    });

    test("the config for Qwik city is added when it is built by ssr mode", () => {
      // not ssr build
      expect(
        (plugin.config as ConfigFunction)(
          {},
          { command: "build", mode: "", ssrBuild: false },
        ),
      ).toStrictEqual({});

      // ssr build
      expect(
        (plugin.config as ConfigFunction)(
          {},
          { command: "build", mode: "", ssrBuild: true },
        ),
      ).toStrictEqual({
        build: {
          rollupOptions: {
            input: [
              "/working-dir/node_modules/.portable/worker.ts",
              "@qwik-city-plan",
            ],
          },
        },
      });
    });

    test("the `@jsxImportSource` hint comment is inserted by `transform` on tsx and jsx file", () => {
      expect(
        (plugin.transform as TransformFunction)("base", "foo.ts"),
      ).toBeUndefined();

      expect((plugin.transform as TransformFunction)("base", "foo.tsx")).toBe(
        `/** @jsxImportSource react */\nbase`,
      );
      expect((plugin.transform as TransformFunction)("base", "foo.jsx")).toBe(
        `/** @jsxImportSource react */\nbase`,
      );
    });
  });
});
