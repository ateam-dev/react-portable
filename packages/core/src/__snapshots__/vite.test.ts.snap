// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`preparePlugin > Install the route file if \`previewify\` is called in the file 1`] = `
{
  "/app/.portable/routes/pfy-sample/index.jsx": "import { qwikify$ } from \\"@builder.io/qwik-react\\";
import { component$ } from \\"@builder.io/qwik\\";
import { routeLoader$ } from \\"@builder.io/qwik-city\\";
import * as Entries from \\"/app/src/index.ts\\";

const Entry = Object.values(Entries).find((module) => {
  if (
    [\\"function\\", \\"object\\"].includes(typeof module) &&
    module !== null &&
    \\"__code\\" in module
  )
    return module.__code === \\"pfy-sample\\";
});

const QComponent = qwikify$(Entry.__forQwik, {
  eagerness: \\"idle\\",
});

const getProps = routeLoader$(async ({ request }) => {
  if (request.method === \\"POST\\") return await request.json();

  return {};
});

export default component$(() => {
  return <QComponent {...getProps().value} />;
});
",
}
`;

exports[`preparePlugin > when \`config\` will be called, \`entry.ssr.jsx\` will be placed 1`] = `
{
  "/app/.portable/entry.ssr.jsx": "import { renderToStream } from \\"@builder.io/qwik/server\\";
import { manifest } from \\"@qwik-client-manifest\\";
import { QwikCityProvider, RouterOutlet } from \\"@builder.io/qwik-city\\";
import { component$ } from \\"@builder.io/qwik\\";

const Root = component$(() => (
  <QwikCityProvider>
    <RouterOutlet />
  </QwikCityProvider>
));

export default function (opts) {
  return renderToStream(<Root />, {
    manifest,
    ...opts,
    containerTagName: \\"rp-fragment\\",
    qwikLoader: { include: \\"always\\" },
  });
}
",
}
`;
